pkg/git/client.go

```diff
@@ -0,0 +1,197 @@
+package git
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"regexp"
+	"sort"
+	"strings"
+
+	"github.com/cli/safeexec"
+)
+
+var remoteRE = regexp.MustCompile(`(.+)\s+(.+)\s+\((push|fetch)\)`)
+
+type Client struct {
+	command  func(ctx context.Context, name string, args ...string) *exec.Cmd
+	dir      string
+	exeErr   error
+	exePath  string
+	stderr   io.Writer
+	stdin    io.Reader
+	stdout   io.Writer
+	workTree string
+}
+
+type ClientOptions struct {
+	Command  func(ctx context.Context, name string, args ...string) *exec.Cmd
+	Dir      string    // will be passed to commands as --git-dir
+	ExePath  string    // will be resolved if none passed
+	Stderr   io.Writer // will use stderr if none specified
+	Stdin    io.Reader // will use stdin if none specified
+	Stdout   io.Writer // will use stdout if none specified
+	WorkTree string    // will be passed to commands as --work-tree
+}
+
+func NewClient(opts *ClientOptions) Client {
```

Comment:
If passing nil options is a valid way to initialize a zero-value Client, then couldn't we make the zero-value `Client{}` struct be useful as a client, and have users do that when there are no options?

And then, if there are some options, users can use `NewClient()` where the options struct would now be a required argument. Just a thought!